File: packages/oxlint-plugin-react-doctor/src/plugin/rules/security-scan/svg-filter-clickjacking-risk.ts
import { defineRule } from "../../utils/define-rule.js";
import { isProductionSourcePath } from "./utils/is-production-source-path.js";
import { scanByPattern } from "./utils/scan-by-pattern.js";

export const svgFilterClickjackingRisk = defineRule({
  id: "svg-filter-clickjacking-risk",
  title: "SVG-filtered iframe clickjacking primitive",
  severity: "warn",
  recommendation:
    "Avoid filtering cross-origin iframes. Use `frame-ancestors` on sensitive pages and keep SVG filters away from embedded privileged UI.",
  // The middle branch's window starts right after `url(#` — an extra `.*`
  // there overlapped the `[\s\S]{0,700}` window and made no-match scans
  // quadratic on long single-line files (seconds at the 2 MB cap).
  scan: scanByPattern({
    shouldScan: (file) => isProductionSourcePath(file.relativePath),
    pattern:
      /<iframe\b[\s\S]{0,700}\bfilter\s*:\s*["']?url\(#|filter\s*:\s*url\(#[\s\S]{0,700}<iframe\b|<fe(?:DisplacementMap|ColorMatrix|Composite|Tile|Morphology)\b[\s\S]{0,700}<iframe\b/i,
    message:
      "An iframe is rendered through an SVG/CSS filter, which can support advanced clickjacking or visual exfiltration tricks.",
  }),
});

File: native/oxlint/scans/svg-filter-clickjacking-risk.rs
use lazy_regex::{Lazy, Regex, lazy_regex};

use super::{
    ScanFinding, get_location_at_index::get_location_at_index, scan_content::ScanContent,
};

const MESSAGE: &str = "An iframe is rendered through an SVG/CSS filter, which can support advanced clickjacking or visual exfiltration tricks.";

static SVG_FILTER_CLICKJACKING_PATTERN: Lazy<Regex> = lazy_regex!(
    r#"(?is)<iframe(?-u:\b)(?:=>|[^>]){0,300}(?-u:\b)filter\s*:\s*[\"']?url\(#|filter\s*:\s*url\(#[\s\S]{0,700}<iframe(?-u:\b)|<fe(?:DisplacementMap|ColorMatrix|Composite|Tile|Morphology)(?-u:\b)[\s\S]{0,700}<iframe(?-u:\b)"#
);

pub fn scan(relative_path: &str, source: &ScanContent<'_>) -> Vec<ScanFinding> {
    if !super::is_production_file_path::is_production_source_path(relative_path) {
        return Vec::new();
    }
    let normalized = source.normalized_scannable(false);
    let Some(found) = SVG_FILTER_CLICKJACKING_PATTERN.find(normalized.as_ref()) else {
        return Vec::new();
    };
    let (line, column) = get_location_at_index(source, normalized.as_ref(), found.start());
    vec![ScanFinding::inherited(MESSAGE, line, column)]
}
